core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java

```diff
@@ -1836,7 +1836,7 @@ public static boolean isValidDecimalValue(@Nullable BigDecimal value, RelDataTyp
     case DECIMAL:
       final int intDigits = value.precision() - value.scale();
       final int maxIntDigits = toType.getPrecision() - toType.getScale();
-      return intDigits <= maxIntDigits;
+      return (intDigits <= maxIntDigits) && (value.scale() <= toType.getScale());
```

Comment:
The question is really about what this function is supposed to implement.
My assumption is that this function decides whether a BigDecimal value can be represented exactly in the specified type. In this case I believe that both conditions need to be satisfied. If the scale is not large enough, the number may need rounding to be represented, so the representation won't be exact.

Reply:
If we don't know what it does and why it exists better not modify it. Like this we avoid accidental changes that are not directly related to this PR.

Reply:
I will check, this may be a leftover that is not needed here, but is needed for https://issues.apache.org/jira/browse/CALCITE-3522. I expect that eventually we will need this code.

Reply:
This change is actually necessary, so I will create a new helper function for it to avoid modifying the existing one.